### PR TITLE
Span merge review

### PR DESCRIPTION
**Pull request by pascalkuthe**
I did a pretty comprehensive review of the SpanIter from helix-editor/helix#3695.
I found a couple of bugs, and some opportunities for performance improvement.
The changes are complex and quite a lot of the so I did not want to suggest them trough the GitHub web UI.
See my comments in helix-editor/helix#3695 for more details

### CODE REVIEW

**Comment by the-mikedavis** [comment]
Sweet, thanks for taking a look at this in such depth! Having another pair of eyes on this should hopefully help iron out those last remaining bugs. It might take me a while to look into the changes properly but a few notes off the top of my head:

* property testing: I thought this might be useful after writing a few unit tests manually; thanks for setting it up 😀
* sorting range ends descending: I had a commit with this change for a while but I dropped it because it exposed more edge cases than sorting ascending. It looks like your changes may have covered the edge cases I found. I will mostly be looking at this while reviewing.

**Comment by pascalkuthe** [comment]
Sorry about the failing tests somehow I made a small mistake while factoring out a function and forgot to save before running cargo test (still used to autosaving from nvim) :D
It should be fixed now 

> sorting range ends descending: I had a commit with this change for a while but I dropped it because it exposed more edge cases than sorting ascending. It looks like your changes may have covered the edge cases I found. I will mostly be looking at this while reviewing.

Ah that makes sense so it was intentional. I wasn't quite sure because the documentation of the `span_iter` function still claimed that spans should be ordered in descending order while elsewhere (and one testcase) used ascending order.

I guess you probably found similar edge cases to me. The edge cases I fixed here actually also occur for ascending order but instead of completely nonsensical spans (end past the start) this leads to subtlety broken highlighting which none of the previous tests would have picked up.
I updated my comment to reflect that (also with some more details). I had that mixed up because I mostly looked at descending order most of the time. 
My fix wouldn't actually work for ascending order (and an alternative fix for ascending order would be more complex I think).

Note that I have removed the `empty_span_at_sublice_start` testcase. This was the minimal reproducible testcase the proptest gave me but its actually a poor example and relied on empty spans to introduce the bug (and therefore doesn't actually reach the edgecase anymore, because empty spans are now ignored)

**Comment by pascalkuthe** [comment]
It seems that dropping empty spans can cause a mismatched number of `HiglightStart`/`HighlightEnd` events. I am not too sure why that is. I don't have time too dig into that right now so I just dropped that one commit since it was pretty trivial

File: helix-core/src/syntax/span.rs [hunk 1]
```diff
@@ -17,9 +21,9 @@ pub struct Span {
 
 impl Ord for Span {
     fn cmp(&self, other: &Self) -> std::cmp::Ordering {
-        // Sort by range: ascending by start and then ascending by end for ties.
+        // Sort by range: ascending by start and then descending by end for ties.
         if self.start == other.start {
-            self.end.cmp(&other.end)
+            other.end.cmp(&self.end)
```
Comment (the-mikedavis) [inline helix-core/src/syntax/span.rs]: minor nit but I prefer `self.end.cmp(&other.end).reverse()` since it's more readable IMO

File: helix-core/src/syntax/span.rs [hunk 2]
```diff
@@ -67,6 +71,79 @@ pub fn span_iter(spans: Vec<Span>) -> impl Iterator<Item = HighlightEvent> {
     }
 }
 
+impl SpanIter {
+    fn start_span(&mut self, span: Span) {
+        use HighlightEvent::*;
```
Comment (the-mikedavis) [inline helix-core/src/syntax/span.rs]: We can just import the enum variants for the whole crate I think, since a span is just a convenience wrapper around them

File: helix-core/src/syntax/span.rs [hunk 3]
```diff
@@ -67,6 +71,79 @@ pub fn span_iter(spans: Vec<Span>) -> impl Iterator<Item = HighlightEvent> {
...
+            ..first_partioned_span
+        };
+        while let Some(span) = self.spans.get(i) {
+            if span <= &intersect_pos {
+                after = Some(i);
+                i += 1;
+            } else {
+                break;
+            }
+        }
```
Comment (the-mikedavis) [inline helix-core/src/syntax/span.rs]: I opted for the `loop { match { Some(range) if .. { .. }, _ => break, } }` pattern since I thought the `while let` version was less clear that we're only looking for exactly one pattern. What do you think, is the `while let` version more idiomatic?

File: helix-core/src/syntax/span.rs [hunk 4]
```diff
@@ -81,125 +158,65 @@ impl Iterator for SpanIter {
...
             }
+            self.start_span(span);
+            self.index += 1;
         }
 
+        // Ensure range ends are sorted in descending orders.
+        // Ranges are sorted by their start instead of their end so the input sorting can't be reused.
+        // The range ends must be sorted in descending order.
+        // So that the ranges that end before the next range can be easily removed.
+        self.range_ends.sort_unstable_by(|lhs, rhs| rhs.cmp(lhs));
```
Comment (the-mikedavis) [inline helix-core/src/syntax/span.rs]: Similar here about the sorting

File: helix-core/src/syntax/span.rs [hunk 5]
```diff
@@ -67,6 +71,79 @@ pub fn span_iter(spans: Vec<Span>) -> impl Iterator<Item = HighlightEvent> {
...
+    fn partition_spans_at(&mut self, intersect: usize) {
+        let first_partioned_span = self.spans[self.index];
+
+        let mut i = self.index;
+        while let Some(span) = self.spans.get_mut(i) {
+            if span.start != self.cursor || span.end < intersect {
+                break;
+            }
+
+            let mut partinoed_span = *span;
```
Comment (the-mikedavis) [inline helix-core/src/syntax/span.rs]: typos for these two vars on "partitioned"

File: helix-core/src/syntax/span.rs [hunk 6]
```diff
@@ -67,6 +71,79 @@ pub fn span_iter(spans: Vec<Span>) -> impl Iterator<Item = HighlightEvent> {
...
+
+        // When spans are subsliced, the span Vec may need to be re-sorted
+        // because the `range.start` may now be greater than some `range.start`
+        // later in the Vec. This is not a classic "sort": we take several
+        // shortcuts to improve the runtime so that the sort may be done in
+        // time linear to the cardinality of the span Vec. Practically speaking
+        // the runtime is even better since we only scan from `self.index` to
+        // the first element of the Vec with a `range.start` after this range.
+        let mut after = None;
+        let intersect_pos = Span {
```
Comment (the-mikedavis) [inline helix-core/src/syntax/span.rs]: I think this would read clearer as `intersect_span`

File: helix-core/src/syntax/span.rs [hunk 7]
```diff
@@ -67,6 +71,79 @@ pub fn span_iter(spans: Vec<Span>) -> impl Iterator<Item = HighlightEvent> {
...
+    fn start_span(&mut self, span: Span) {
+        use HighlightEvent::*;
+
+        debug_assert!(span.start <= span.end);
+        self.event_queue
+            .push_back(HighlightStart(Highlight(span.scope)));
+        self.range_ends.push(span.end);
+    }
+
+    fn process_range_end(&mut self, end: usize) -> HighlightEvent {
```
Comment (the-mikedavis) [inline helix-core/src/syntax/span.rs]: another minor nit: I think this would be clearer as `emit_range_end` since it is either queueing or directly returning the ending event

**Review by the-mikedavis — COMMENTED** [review]
I was only able to look at this superficially so far but I'm liking what I see 😀

**Comment by the-mikedavis** [comment]
From the upstream PR review:

> I haven't looked at the merge implementation yet, I think that will also be a good fit for proptesting

I'm much more confident about that implementation: the procedure is much simpler - basically the same problem as merging two lists which are already sorted - I haven't found any bugs in it yet (compared to `span_iter` which has been quite a headache), and it's essentially the same as the original implementation of `Merge`.

That being said, I agree it's a good place for more proptesting 👍

File: helix-core/src/syntax/span.rs (hunk 7 above)
Comment (pascalkuthe) [reply helix-core/src/syntax/span.rs]: fixed in 20302313b5d274d231c8c2e505a5306601c0173d

File: helix-core/src/syntax/span.rs (hunk 5 above)
Comment (pascalkuthe) [reply helix-core/src/syntax/span.rs]: fixed in 20302313b5d274d231c8c2e505a5306601c0173d

File: helix-core/src/syntax/span.rs (hunk 4 above)
Comment (pascalkuthe) [reply helix-core/src/syntax/span.rs]: fixed in 20302313b5d274d231c8c2e505a5306601c0173d

File: helix-core/src/syntax/span.rs (hunk 2 above)
Comment (pascalkuthe) [reply helix-core/src/syntax/span.rs]: fixed in 20302313b5d274d231c8c2e505a5306601c0173d

File: helix-core/src/syntax/span.rs (hunk 1 above)
Comment (pascalkuthe) [reply helix-core/src/syntax/span.rs]: fixed in 20302313b5d274d231c8c2e505a5306601c0173d

File: helix-core/src/syntax/span.rs (hunk 6 above)
Comment (pascalkuthe) [reply helix-core/src/syntax/span.rs]: fixed in 20302313b5d274d231c8c2e505a5306601c0173d.

This made me notice that the term `range` is kind of used randomly instead of `span`. I think it might be nice to consistently use the term `span` instead. I did that in 7de6f2a8ab7f47342d5ab017fca87cf05b9a7f5b and also expanded on the documentation of `partition_spans_at`. Do you think the renaming is better or should I revert that commit (except for the documentation)?

File: helix-core/src/syntax/span.rs (hunk 3 above)
Comment (pascalkuthe) [reply helix-core/src/syntax/span.rs]: I personally like the while loop better, because it looks less cluttered.
I personally always prefer `if` to `match` if there is just one case that I actually care about.
Its also closer to my ideal solution which would be a let_chain whenever that becomes available on a rust version we can use: `while let Some(span) = self.spans.get(i) && span <= &intersect_pos `

The `while` loop is just a standard vec iterator so when scanning the code its immediately) obvious what it does and I can focus on the content of the loop (where anything interesting is happening).

However thinking about this more this can actually be solved with an iterator which makes it much clearer.
This also inspired me to rename some variables to make the behaviour of `partition_spans_at` much clearer.
These changes are in 30563c4dd5c5d727c41af39abf665ebee2065ba4. Let me know what you think

**Comment by pascalkuthe** [comment]
> It seems that dropping empty spans can cause a mismatched number of `HiglightStart`/`HighlightEnd` events. I am not too sure why that is. I don't have time too dig into that right now so I just dropped that one commit since it was pretty trivial

I have fixed this now in 8f995d6dc3a4e061411d7700e8d43d172bf59ff5. I simply dropped the empty spans too late where they already had some influence on the state of the iterator and ignoring them left the iterator in an invalid state.

I now drop the right at the start of the iterator before any state has been modified.

Protesting is really invaluable for this specific task. Whenever I make an oversight it is always caught (even tough none of the existing testcases found it). While I am not 100% sure about this I am reasonable sure that the proptest can reach 100% coverage on all permutations of the algorithm. I will look into improving the prop testing even further (and the applying it to merge)

**Comment by pascalkuthe** [comment]
Another update from me (sorry for the spam).

I have significantly improved the proptesting infrastructure.
I have created a datastructure that allows me to represent the highlight of each character as a bitset.
This datastructure can be easily constructed from both highlight events and spans directly.
This allows me to compare the highlighting produced by the span_iter to that of the original evenets.

This has turned up two minor bugs (which are just extensions of what I already worked on):
* empty spans must always be removed because they can actually mess up the highlight of other spans, in particular span partioning can create new empty spans by moving the start of a span to its end
* previously the spans vector was resorted using `rotate_left` this was a very nice optimization but sadly only works if the spans were only sorted by their start. I already noticed during my initial review that this is not correct because span ends must also stay in descending order. However my suggested fix was wrong: I simply also rotate the spans that had a larger end to the front. However the span ends are not actually sorted (or rather there are two distincs sorted lists) and therefore if that special case occurs, a (very small) part of the spans vec needs a full call to `sort_unstable`

I have pushed the fixes. Right now I am letting my PC run a billion proptestcases to ensure there are no edgecases left but I am already at a couple hundred million so I doubt any more will turn up. The proptests now cover the entire output of the algorithm so we have realistically 100% coverage once that tests concludes.

I will start looking at the merge implementation now.

**Comment by pascalkuthe** [comment]
(closed in favor of helix-editor/helix#4042)
